shopinvader_address/models/res_partner.py

```diff
@@ -0,0 +1,162 @@
+                    address_id=self.id,
+                )
+            )
+
+        # if billing addresses is already used, it is not possible to modify it
+        if self._shopinvader_billing_address_already_used():
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+        self.write(vals)
+
+        return self
+
+    # ---Shipping ---
+
+    def _get_shopinvader_shipping_addresses(
+        self, address_id: int = None
+    ) -> "ResPartner":
+        self.ensure_one()
+        domain = [("type", "=", "delivery"), ("parent_id", "=", self.id)]
+
+        if address_id is not None:
+            domain.append(("id", "=", address_id))
+
+        return self.env["res.partner"].search(domain)
+
+    def _create_shopinvader_shipping_address(self, vals: dict) -> "ResPartner":
+        self.ensure_one()
+        vals.update(
+            {
+                "parent_id": self.id,
+                "type": "delivery",
+            }
+        )
+        return self.env["res.partner"].create(vals)
+
+    def _update_shopinvader_shipping_address(
+        self, vals: dict, address_id: int
+    ) -> "ResPartner":
+
+        if any(key in vals for key in ("parent_id", "type")):
+            raise UserError(
+                _(
+                    "parent_id and type cannot be modified on"
+                    " shopinvader shipping address, id: %(address_id)d",
+                    address_id=address_id,
+                )
+            )
+
+        self.ensure_one()
+        address = self._get_shopinvader_shipping_addresses(address_id)
+        if not address:
+            raise MissingError(
+                _("No address found, id: %(address_id)d", address_id=address_id)
+            )
+
+        # if shipping addresses is already used, it is not possible to modify it
+        if address._shopinvader_shipping_address_already_used():
+            raise UserError(
+                _(
+                    "Can not update Shipping address(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+        # update_address
+        address.write(vals)
+        return address
+
+    def _delete_shopinvader_shipping_address(self, address_id: int) -> None:
+        """
+        Delete of shopinvader shipping addresses will result to an archive
+        """
+        address = self._get_shopinvader_shipping_addresses(address_id)
+        if address:
+            if address._shopinvader_shipping_address_already_used():
+                raise UserError(
+                    _(
+                        "Can not delete Shipping address(%(address_id)d)"
+                        "because it is already used on confirmed sale order",
```

Comment:
The fact that this error message is repeated here indicates that we should refactor.